src/python/py_generate_pipeline.cpp

```diff
@@ -23,88 +27,276 @@ PYBIND11_MAKE_OPAQUE(std::vector<float>);
+            return_value[key] = py_object_to_any_map(value);
+        } else {
+            return_value[key] = py_object_to_any(value);
+        }
+    }
+    return return_value;
 }
 
-OpaqueDecodedResults call_with_config(LLMPipeline& pipe, const std::vector<std::string>& text, const GenerationConfig& config, const StreamerVariant& streamer) {
-    return OpaqueDecodedResults{pipe.generate(text, config, streamer)};
+std::map<std::string, ov::Any> properties_to_any_map(const std::map<std::string, py::object>& properties) {
+    std::map<std::string, ov::Any> properties_to_cpp;
+    for (const auto& property : properties) {
+        properties_to_cpp[property.first] = py_object_to_any(property.second);
+    }
+    return properties_to_cpp;
 }
 
-OpaqueDecodedResults call_with_kwargs(LLMPipeline& pipeline, const std::vector<std::string>& texts, const py::kwargs& kwargs) {
-    GenerationConfig config = pipeline.get_generation_config();
-    update_config_from_kwargs(config, kwargs);
-    return call_with_config(pipeline, texts, config, kwargs.contains("streamer") ? kwargs["streamer"].cast<StreamerVariant>() : std::monostate());
+py::list handle_utf8_results(const std::vector<std::string>& decoded_res) {
+    // pybind11 decodes strings similar to Pythons's
+    // bytes.decode('utf-8'). It raises if the decoding fails.
+    // generate() may return incomplete Unicode points if max_new_tokens
+    // was reached. Replace such points with � instead of raising an exception
+    py::list res;
+    for (const auto s: decoded_res) {
+        PyObject* py_s = PyUnicode_DecodeUTF8(s.data(), s.length(), "replace");
+        res.append(py_s);
+    }
+    return res;
 }
 
-OpaqueDecodedResults call_with_kwargs(LLMPipeline& pipeline, const std::string& text, const py::kwargs& kwargs) {
-    // Create a new GenerationConfig instance and initialize from kwargs
-    GenerationConfig config = pipeline.get_generation_config();
-    update_config_from_kwargs(config, kwargs);
-    return call_with_config(pipeline, text, config, kwargs.contains("streamer") ? kwargs["streamer"].cast<StreamerVariant>() : std::monostate());
+py::object call_common_generate(
+    LLMPipeline& pipe, 
+    const std::variant<EncodedInputs, StringInputs>& inputs, 
+    const OptionalGenerationConfig& config, 
+    const StreamerVariant& streamer, 
+    const py::kwargs& kwargs
+) {
+    auto updated_config = update_config_from_kwargs(config, kwargs);
+
+    if (auto data = std::get_if<EncodedInputs>(&inputs)) {
+        return py::cast(pipe.generate(*data, updated_config, streamer));
+    } else if (auto data = std::get_if<StringInputs>(&inputs)) {
+        DecodedResults res = pipe.generate(*data, updated_config, streamer);
+        // If input was a string return a single string otherwise return DecodedResults.
+        if (std::get_if<std::string>(&*data) && updated_config.num_return_sequences == 1) {
+            return handle_utf8_results(res.texts)[0];
```

Comment:
If i feed a single string i expect to get a string at output. So pythons behavior is better. In older version we had a c++ overload for that.

But currently we always return `DecodedResults` because we don't want to add one more overload for `generate(std::string, ...)` when we already have so many of overloads.

Reply:
The number of overloads isn't a problem by itself. The more specifications you have the more options you cover for the user. And it definitely can't be an argument for introducing misalignment. The reason why it it's acceptable for now is the lack of time.